client/state/query-client-ssr.ts

```diff
@@ -1,19 +1,34 @@
-import { QueryClient, QueryCache, dehydrate } from 'react-query';
+import { QueryClient, QueryCache, dehydrate, hashQueryKey, TQueryKey } from 'react-query';
 import { MAX_AGE, BASE_STALE_TIME } from 'calypso/state/initial-state';
 const sharedCache = new QueryCache();
 
+class QueryClientSSR extends QueryClient {
+	fetchedQueryKeys: { [ hash: string ]: boolean } = {};
+
+	fetchQuery( queryKey: TQueryKey, ...args: any[] ) {
+		const queryHash = hashQueryKey( queryKey );
```

Comment:
Does this really work? If I call `prefetchQuery()`, it will internally merge all arguments (key, fn, options) into one `options` object and call `this.fetchQuery` with a single object argument. And then your function will be hashing an object that's not a key?

There is also `fetchInfiniteQuery` which merges its arguments the same way.

Also, the user can pass a custom `options.queryKeyHashFn` which will be used instead of `hashQueryKey`.

Maybe it's better to create custom functions like:
```js
fetchQuery( queryClient, key, fn, options );
fetchInfiniteQuery( queryClient, key, fn, options );
```
without subclassing `QueryClient`, and using them exclusively in SSR handlers.

Reply:
>  If I call prefetchQuery(), it will internally merge all arguments (key, fn, options) into one options object and call this.fetchQuery with a single object argument.

Thanks for catching it. I didn't realize this was happening. 
I only see this happening for [fetchInfiniteQuery](https://github.com/TanStack/query/blob/v3.32.1/src/core/queryClient.ts#L453) which is not called yet for the Plugins Landing Page.
We should probably check for the `queryKey` on the possible object. On the other side, we don't really need to have this `queryHash` match the one inside `QueryClient`. We just need it to match the hashes we generate in the dehydration step (which is not happening currently).

> Maybe it's better to create custom functions like ... without subclassing QueryClient, and using them exclusively in SSR handlers.

The functions should allow us to not check for the possible passed object, but I think we'll still need to subclass `QueryClient` in order to store the fetched queries so that they can be used in the rendering/dehydration step.
Alternatively, we could add another property to the `context` object, but I think encapsulating this data inside the `QueryClient` instance is the cleaner approach.